README.md
# Google Tag Manager for SharePoint

SPFx Extension to implement Google Tag Manager solution.

This solution provide the solution with the latest version of SharePoint Framework and all scripts to install and deploy the solution to your environment.

## PowerShell Setup (Windows only)

### Prerequisites

1. A Google Tag Manager Tracking ID
2. Site Collection App Catalog or Tenant App Catalog
3. Install [PnP PowerShell Module](https://github.com/SharePoint/PnP-PowerShell/releases)
4. Login to your target SharePoint Site Collection (more information about [Connect-PnPOnline](https://docs.microsoft.com/en-us/powershell/module/sharepoint-pnp/connect-pnponline))
    ```powershell
    Connect-PnPOnline -Url https://contoso.sharepoint.com/sites/target-site
    ```

### Setup

| Argument | Required | Default value | Description |
|----------|----------|---------------|-------------|
|`-siteUrl`|yes|`undefined`|URL of the site to provision and/or enable the extension|
|`-trackingID`|yes|`undefined`|Google Tag Manager Tracking ID|
|`-tenantSolutionDeployment`|no|`false`|Set `true`, to deploy the solution package to the whole tenant. If not specified, the package will be deployed to the current Site App Catalog|
|`-skipCustomAction`|no|`false`|Don't enable the custom action to the target site|
|`-checkPoint`|no|`0`|Check point from which to resume executing the setup script|
|`-appId`|no|`undefined`|Product ID of the App from the App Catalog. Use only at checkpoint `200`|

> Note: to get help from PowerShell cmdlet, type : `Get-Help .\setup.ps1`

#### Examples

```powershell
.\setup.ps1 -siteUrl https://contoso.sharepoint.com/sites/target-site -trackingID 'GTM-UMA0000'
```
In this example, deploy the solution package to the Site Collection App Catalog

```powershell
.\setup.ps1 -siteUrl https://contoso.sharepoint.com/sites/target-site -trackingID 'GTM-UMA0000' -tenantSolutionDeployment
```
In this example, deploy the solution package to the Tenant App Catalog

```powershell
.\setup.ps1 -siteUrl https://contoso.sharepoint.com/sites/target-site -trackingID 'GTM-UMA0000' -checkPoint 300
```
In this example, enable the custom action the target site collection only

```powershell
.\setup.ps1 -siteUrl https://contoso.sharepoint.com/sites/target-site -trackingID 'GTM-UMA0000' -skipCustomAction
```
In this example, deploy and install the package only (the custom action will be not enabled)

## Bash Setup (any platform)

### Prerequisites

1. A Google Tag Manager Tracking ID
2. Site Collection App Catalog or Tenant App Catalog
3. Install [Office365 CLI](https://pnp.github.io/office365-cli/user-guide/installing-cli)
    ```bash
    npm i -g @pnp/office365-cli
    ```
4. Execute
    * login to your target SharePoint Site Collection (more information about [spo login](https://pnp.github.io/office365-cli/cmd/spo/login))
    ```bash
    o365 spo connect https://contoso.sharepoint.com/sites/target-site
    ```
    * allow execution setup script
    ```bash
    chmod +x ./setup.sh
    ```
    * execute setup script

### Setup

| Argument | Required | Default value | Description |
|----------|----------|---------------|-------------|
|`-s`,`--siteUrl`|yes|`undefined`|URL of the site to provision and/or enable the extension|
|`--trackingID`|yes|`undefined`|Google Tag Manager Tracking ID|
|`--tenantSolutionDeployment`|no|`false`|Set `true`, to deploy the solution package to the whole tenant. If not specified, the package will be deployed to the current Site App Catalog|
|`--skipCustomAction`|no|`false`|Don't enable the custom action to the target site|
|`--checkPoint`|no|`0`|Check point from which to resume executing the setup script|
|`--appId`|no|`undefined`|Product ID of the App from the App Catalog. Use only at checkpoint `200`|
|`--verbose`|no|`false`|Runs setup with verbose logging|
|`--help`|no|`false`|Output usage information|

#### Examples

```bash
./setup.sh --siteUrl https://contoso.sharepoint.com/sites/target-site --trackingID 'GTM-UMA0000'
```
In this example, deploy the solution package to the Site Collection App Catalog

```bash
./setup.sh --siteUrl https://contoso.sharepoint.com/sites/target-site --trackingID 'GTM-UMA0000' --tenantSolutionDeployment
```
In this example, deploy the solution package to the Tenant App Catalog

```bash
./setup.sh --siteUrl https://contoso.sharepoint.com/sites/target-site --trackingID 'GTM-UMA0000' --checkPoint 300
```
In this example, enable the custom action the target site collection only

```bash
./setup.sh --siteUrl https://contoso.sharepoint.com/sites/target-site --trackingID 'GTM-UMA0000' --skipCustomAction
```
In this example, deploy and install the package only (the custom action will be not enabled)


CHANGELOG.md
## [0.0.3](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/compare/v0.0.2...v0.0.3) (2020-01-20)


### Bug Fixes

* partial loading page ([0d58f79](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/0d58f796b7144e8129360ec79350c6820d37312d))
* **setup:** rename functions file for bash install ([7224e0f](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/7224e0fcb0cc58648c67fc28e1e601d6abbbd65a))


### Features

* remove auto-deploy custom action to avoid doublon ([e235d02](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/e235d02eac70583d085921d41eb7897e236ac598))
* update version ([10f028b](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/10f028b7fbaaff558fed187bf8bc90730933e920))



## [0.0.2](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/compare/v0.0.1...v0.0.2) (2020-01-14)


### Features

* **doc:** update documentation and rename clientsiteinstance ([7ca0bd2](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/7ca0bd29105105a26c77c476be42686eb99cd34f))



## 0.0.1 (2020-01-14)


### Bug Fixes

* CI format source file ([e02cb89](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/e02cb8969e523b95664d205e23e9e88d3b711de0))


### Features

* **CI:** init DevOps script ([018b21d](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/018b21d225497904fe2baa5609cd86c9112bb2fc))
* Add deployment scripts ([7fdac5b](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/7fdac5b368fb4f34811c78e01dc4a5e68f1de0bb))
* first version of the solution ([01bc733](https://gitlab.lsonline.fr/SharePoint/sp-dev-fx-webparts/gtm/commit/01bc733699da29dacdbd50c8d0fa3117fdb96522))





setup.sh
#!/usr/bin/env bash

# helper functions
. ./_functions.sh

# default args values
siteUrl=
trackingID=""
tenantSolutionDeployment=false
verbose=false
skipCustomAction=false
checkPoint=0
appId=

# script arguments
while [ $# -gt 0 ]; do
  case $1 in
    -s|--siteUrl)
      shift
      siteUrl=$1
      ;;
    --trackingID)
      shift
      trackingID=$1
      ;;
    --tenantSolutionDeployment)
      tenantSolutionDeployment=true
      ;;
    --skipCustomAction)
      skipCustomAction=true
      ;;
    --checkPoint)
      shift
      checkPoint=$1
      ;;
    --appId)
      shift
      appId=$1
      ;;
    --verbose)
      verbose=true
      ;;
    -h|--help)
      help
      exit
      ;;
    *)
      error "Invalid argument $1"
      exit 1
  esac
  shift
done

if [ -z "$siteUrl" ]; then
  error "Please specify site collection URL"
  echo
  help
  exit 1
fi

if [ -z "$trackingID" ]; then
  error "Please specify Tracking ID"
  echo
  help
  exit 1
fi

if [ "$tenantSolutionDeployment" = true ]; then
  # First, find the Tenant App Catalog
  msg "Retrieving tenant app catalog URL...\n"

  appCatalogUrl=$(o365 spo tenant appcatalogurl get)
  if [ -z "$appCatalogUrl" ]; then
    error "Couldn't retrieve Tenant App Catalog"
    exit 1
  fi
  if [ "$verbose" = true ]; then
    msg "Tenant App Catalog was find.\n"
  fi

  msg "Deploying the Google Tag Manager Package to the Tenant AppCatalog...\n"

  if (( $checkPoint < 100 )); then
    # Push package to the Tenant AppCatalog
    if [ "$verbose" = true ]; then
      msg "Adding the Google Tag Manager Package to the Tenant AppCatalog...\n"
    fi
    appId=$(o365 spo app add --filePath ./gtm-for-sharepoint.sppkg)

    if [ "$verbose" = true ]; then
      msg "App ID: $appId...\n"
    fi

    checkPoint=100
  fi
  if (( $checkPoint < 200 )); then
    # Deploy solution from the Tenant AppCatalog
    if [ "$verbose" = true ]; then
      msg "Deploying the Google Tag Manager Package...\n"
    fi
    o365 spo app deploy --name Google Tag Manager.sppkg --skipFeatureDeployment

    checkPoint=200
  fi
  if (( $checkPoint < 300 )); then
    # Install solution to site collection
    if [ "$verbose" = true ]; then
      msg "Installing the Google Tag Manager Package with ID: $appId...\n"
    fi
    o365 spo app install --id $appId --siteUrl $siteUrl

    checkPoint=300
  fi
else
  if (( $checkPoint < 100 )); then
    msg "Deploying the Google Tag Manager Package to the Site AppCatalog...\n"
    # Push package to the site AppCatalog
    if [ "$verbose" = true ]; then
      msg "Adding the Google Tag Manager Package to the Site AppCatalog $siteUrl...\n"
    fi
    appId=$(o365 spo app add --filePath ./gtm-for-sharepoint.sppkg --scope sitecollection --appCatalogUrl $siteUrl)

    if [ "$verbose" = true ]; then
      msg "App ID: $appId...\n"
    fi
    
    checkPoint=100
  fi
  if (( $checkPoint < 200 )); then
    # Deploy solution from the site AppCatalog
    if [ "$verbose" = true ]; then
      msg "Deploying the Google Tag Manager Package...\n"
    fi
    o365 spo app deploy --name gtm-for-sharepoint.sppkg --scope sitecollection --appCatalogUrl $siteUrl

    checkPoint=200
  fi
  if (( $checkPoint < 300 )); then
    # Install solution to site collection
    if [ "$verbose" = true ]; then
      msg "Installing the Google Tag Manager Package with ID : $appId...\n"
    fi
    o365 spo app install --id $appId --siteUrl $siteUrl --scope sitecollection

    checkPoint=300
  fi
fi

if [ "$skipCustomAction" = false ]; then
  msg "Enabling the Google Tag Manager extension...\n"

  # Add Custom action to site collection
  str="'{\"trackingID\":\"$trackingID\"}'"

  o365 spo customaction add --url $siteUrl --clientSideComponentId 87dac560-8541-4aef-b094-39b0a0d73985 --name 'Google Tag Manager for SharePoint' --title 'Google Tag Manager for SharePoint' --location 'ClientSideExtension.ApplicationCustomizer' --scope Site -p $str
fi
